interfaces.go

```diff
@@ -4,28 +4,54 @@
 
 package optimize
 
-// LinesearchMethod is a type that can perform a line search. Typically, these
-// methods will not be called by the user directly, as they will be called by
-// a Linesearch struct.
-type LinesearchMethod interface {
-	// Init initializes the linesearch method. LinesearchLocation contains the
-	// function information at step == 0, and step contains the first step length
-	// as specified by the NextDirectioner.
-	Init(loc LinesearchLocation, step float64) EvaluationType
+// A Method can optimize an objective function.
+type Method interface {
+	// Init initializes the method and stores the first location to evaluate
+	// in xNext.
+	Init(loc *Location, xNext []float64) (EvaluationType, IterationType, error)
+
+	// Iterate performs one iteration of the method and stores the next
+	// location to evaluate in xNext.
+	Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error)
+
+	// Needs specifies information about the objective function needed by the
+	// optimizer beyond just the function value. The information is used
+	// internally for initialization and must match evaluation types returned
+	// by Init() and Iterate() during the optimization process.
+	Needs() struct {
+		Gradient bool
+		Hessian  bool
+	}
+}
+
+// Linesearch is a type that can perform a line search. It tries to find an
+// (approximate) minimum of the objective function along the search direction
+// dir_k starting at the most recent location x_k, i.e., it tries to minimize a
+// function
+//  φ(step) := f(x_k + step * dir_k), step > 0.
+// Typically, these methods will not be called by the user directly, as they
+// will be called by the LinesearchHelper struct.
+type Linesearch interface {
```

Comment:
I think this would be nicer called "Linesearcher", meaning something that can perform a linesearch. The interface doesn't fall into the strict go rules, but I think the code reads a little more clearly with the "er".